### PR TITLE
Restore compatibility with Cassandra 1.2 (fixes #579)

### LINKED ISSUE #579: 1.3.0 on C* 1.2.19: cannot load cluster state (`nodes_status` remains null)
Hi,

I recently upgraded reaper from 1.2.2 to 1.3.0 and since then the UI is not able to load the cluster state anymore, it keeps spinning (it doesn't show neither the disk usage neither the nodes):

spinning

Looking at the ajax calls made I can see that the endpoint `cluster/staging-cluster` returns a json object with `nodes_status` set to `null`:

image

In `cassandra-reaper.log` I can see that the following exception happens frequently:

```
Caused by: javax.management.InstanceNotFoundException: org.apache.cassandra.net:type=StreamManager
```

Note that I haven't checked the code (yet) so I don't know if this may be related or not.

Does it means that reaper 1.3.0 is not compatible with Cassandra 1.2 anymore?

Cheers,
Reynald

### PR DESCRIPTION
**Pull request by rborer**
the StreamManagerMBean mbean has been added to Cassandra 2.0 only. This PR attempts to not connect to this nonexistent mbean on clusters below 2.0.

I successfully ran unit tests but please note that I haven't tested this change against a real cluster. So it needs a strong review :wink:

### CODE REVIEW

**Review by adejanovski — CHANGES_REQUESTED** [review]
Hi @rborer ,

first of all, thanks for submitting a fix for your issue! It's great to have users that submit fixes on top of issues 👍 

I just have a minor concern with the use of `null` and would rather have an Optional used.

Other than that, I'm +1 to merge the change.

File: src/server/src/main/java/io/cassandrareaper/jmx/JmxProxyImpl.java [hunk 1]
```diff
@@ -218,11 +218,19 @@ private static JmxProxy connect(
...
         ssProxy = JMX.newMBeanProxy(mbeanServerConn, ObjectNames.STORAGE_SERVICE, StorageServiceMBean20.class);
       }
 
+      final StreamManagerMBean smProxy;
+      // StreamManagerMbean is only available since Cassandra 2.0
+      if (cassandraVersion.startsWith("1.")) {
+        smProxy = null;
+      } else {
+        smProxy = JMX.newMBeanProxy(mbeanServerConn, ObjectNames.STREAM_MANAGER, StreamManagerMBean.class);
+      }
```
[the remark this answers is not in the document]
Comment (rborer) [reply src/server/src/main/java/io/cassandrareaper/jmx/JmxProxyImpl.java:232]: Sure. I'm so much used to `null` that I didn't even consider `Optional` usage here. I'll update the code accordingly.

[the remark this answers is not in the document]
Comment (rborer) [reply src/server/src/main/java/io/cassandrareaper/jmx/JmxProxyImpl.java:232]: @adejanovski I just pushed a code change, is that what you meant?

I'm also wondering if I should modify the `JmxProxyImpl` constructor signature to accept an `Optional<StreamManagerMBean>` instead? My IDE (IntelliJ) tend to say it's a bad practice as it shows the following code inspection warning:

> Optional was designed to provide a limited mechanism for library method return types where there needed to be a clear way to represent "no result". Using a field with type java.util.Optional is also problematic if the class needs to be Serializable, which java.util.Optional is not.

[the remark this answers is not in the document]
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/jmx/JmxProxyImpl.java:232]: Yes, I would pass an optional instead and modify the signature. We do it a lot in Reaper's codebase, and we don't need the classes to be Serializable here.

[the remark this answers is not in the document]
Comment (rborer) [reply src/server/src/main/java/io/cassandrareaper/jmx/JmxProxyImpl.java:232]: Alright, it's done. Let me know if you need anything else to be modified.

**Review by adejanovski — APPROVED** [review]
LGTM

**Comment by rborer** [comment]
@adejanovski and thanks for this great project, it's extremely useful (and I'm looking forward to `cstar` integration :wink: ).

**Pull request merged**